src/app/student/_components/ExhibitorFilters.tsx

```diff
@@ -0,0 +1,112 @@
+import MultiSelect from "@/app/student/_components/MultiSelect"
+import { Exhibitor } from "@/components/shared/hooks/api/useExhibitors"
+import { Input } from "@/components/ui/input"
+import { useRef, useState } from "react"
+
+// Filtering assumptions:
+// - selecting multiple options for a filter gives the union (not intersection) of those options
+// - selecting no options is the same as selecting all options
+// - filters are ignored when using the search bar
+
+export type FilterKey = "employments" | "industries"
+export type FilterItem = Exhibitor[FilterKey][number]
+
+export type Filter = {
+  key: FilterKey
+  items: FilterItem[]
+  selected: FilterItem[]
+  label: string
+}
+
+function satisfiesFilter(exhibitor: Exhibitor, filter: Filter) {
+  if (filter.selected.length === 0) return true
+  return exhibitor[filter.key].some(x =>
+    filter.selected.some(y => x.id === y.id)
+  )
+}
+
+function applyFilters(exhibitors: Exhibitor[], filters: Filter[]) {
+  return exhibitors.filter(e => filters.every(f => satisfiesFilter(e, f)))
+}
+
+function filterBySearch(exhibitors: Exhibitor[], text: string) {
+  return exhibitors.filter(e =>
+    e.name.toLowerCase().includes(text.toLowerCase())
+  )
+}
+
+// Gets all the possible options by looping over each exhibitor
+function getAllFilterOptions(
+  key: FilterKey,
+  exhibitors: Exhibitor[]
+): FilterItem[] {
+  const distinct = new Map<FilterItem["id"], FilterItem>()
+  exhibitors.forEach(e => {
+    e[key].forEach(item => distinct.set(item.id, item))
```

Comment:
I added an array check to make it safer. The key is defined at the beginning of the file and is limited to only `employments` or `industries`.

Reply:
Could be wrong but I don't think the runtime check is necessary since passing anything other than "employments" or "industries" will be a type error, and Exhibitor["employments" | "industries"] is known by typescript to be an array, so even adding non-array keys to the definition of FilterKey or changing the definition of Exhibitor would result in a type error in the filters file. So I ended removing the check again, but let me know if you think this was incorrect

Reply:
I agree the checking might be not necessary here, since typescript already has the type check, and we already limited the key. Just to intentionally prevent the array to be null. But in the schema definition we don’t allow null value in Exhibitor object. So I think it is safe enough to not have the checking.